netket/sampler/base.py

```diff
@@ -60,14 +60,55 @@ class Sampler(abc.ABC):
     hilbert: AbstractHilbert = struct.field(pytree_node=False)
     """Hilbert space to be sampled."""
 
-    n_chains: int = struct.field(pytree_node=False, default=16)
-    """Number of batches along the chain"""
+    n_chains_per_rank: int = struct.field(pytree_node=False)
+    """Number of independent chains on every MPI rank."""
 
     machine_pow: int = struct.field(default=2)
-    """Exponent of the pdf sampled"""
+    """Exponent of the pdf sampled."""
 
-    dtype: type = struct.field(pytree_node=False, default=np.float64)
-    """DType of the states returned."""
+    dtype: Dtype = struct.field(pytree_node=False, default=np.float64)
+    """DType of the computed samples."""
+
+    def __init__(
+        self,
+        hilbert: AbstractHilbert,
+        n_chains: Optional[int] = None,
+        *,
+        n_chains_per_rank: Optional[int] = None,
+        machine_pow: int = 2,
+        dtype: Dtype = np.float64,
+    ):
+        """
+        Construct a Monte Carlo sampler.
+
+        Args:
+            hilbert: Hilbert space to be sampled
+            n_chains: The total number of independent chains across all MPI ranks
+            n_chains_per_rank: Number of independent chains on every MPI rank
+            machine_pow: Exponent of the pdf sampled
+            dtype: DType of the computed samples
+        """
+
+        # Default number of total chains
+        if n_chains is None and n_chains_per_rank is None:
+            n_chains = 16
+
+        if n_chains is not None:
+            if n_chains_per_rank is not None:
+                raise ValueError(
+                    "Only one between n_chains and n_chains_per_rank can be specified."
+                )
+            else:
+                n_chains_per_rank = n_chains_per_rank = max(
+                    int(np.ceil(n_chains / mpi.n_nodes)), 1
+                )
+                if mpi.n_nodes > 1 and mpi.rank == 0:
+                    print(f"Using {n_chains_per_rank} chains per rank")
```

Comment:
That seems like a debug-type print that should be removed in production.

Reply:
My idea was to leave it there instead.
And only print it on rank 0.

Reply:
Hm, I feel it's not the best practice for library code to write to `stdout` for what is essentially debug output and I think we've removed almost all of than in v3 (we mostly communicate with users via `Warnings`s now :grin:). Especially since it's not too uncommon to recreate sampler several times in a script (at least I've done things like that regularly and I have tons of logfiles where NetKet v2 is printing all kinds of unnecessary information several times because I'm creating several objects.)

I guess the clean way is to set up proper `logging`, but that's of course out of scope of this PR.

Reply:
yeah i thought about that.
I'll throw a warning then

Reply:
I'd throw a warning if `n_chains` does not match what the user specified because of the rounding. Because that can definitely trip people up.

Reply:
is it ok now?